src/extensions/registries/page-registry.ts

```diff
@@ -1,93 +1,95 @@
    * When not provided, first registered page without "id" would be used for page-menus without target.pageId for same extension
    */
   id?: string;
+  components: PageComponents;
   /**
-   * Strict route matching to provided page-id, read also: https://reactrouter.com/web/api/NavLink/exact-bool
-   * In case when more than one page registered at same extension "pageId" is required to identify different pages,
-   * It might be useful to provide `exact: true` in some cases to avoid overlapping routes.
-   * Without {exact:true} second page never matches since first page-id/route already includes partial route.
-   * @example const pages = [
-   *  {id: "/users", exact: true},
-   *  {id: "/users/:userId?"}
-   * ]
-   * Pro-tip: registering pages in opposite order will make same effect without "exact".
+   * Registered page params.
+   * Used to generate page url when provided in getExtensionPageUrl()-helper.
    */
-  exact?: boolean;
-  components: PageComponents;
-}
-
-export interface RegisteredPage extends PageRegistration {
-  extensionId: string; // required for compiling registered page to url with page-menu-target to compare
-  routePath: string; // full route-path to registered extension page
+  params?: UrlParam[];
 }
 
 export interface PageComponents {
   Page: React.ComponentType<any>;
 }
 
-export function getExtensionPageUrl<P extends object>({ extensionId, pageId = "", params }: PageMenuTarget<P>): string {
-  const extensionBaseUrl = compile(`/extension/:name`)({
-    name: sanitizeExtensionName(extensionId), // compile only with extension-id first and define base path
-  });
-  const extPageRoutePath = path.posix.join(extensionBaseUrl, pageId);
+export interface PageTarget<P = {}> {
+  extensionId?: string;
+  pageId?: string;
+  params?: Record<string, any | any[]> & P; // default target page params
+}
+
+export interface RegisteredPage extends PageRegistration {
+  extensionId: string;
+  url: string; // registered extension's page URL (without page params)
+}
+
+export function getExtensionPageUrl(target: PageTarget): string {
+  const { extensionId, pageId = "", params: targetParams = {} } = target;
+  let stringifiedParams = "";
+
+  // stringify params to matched target page
+  const page = globalPageRegistry.getByPageTarget(target) || clusterPageRegistry.getByPageTarget(target);
+
+  if (page?.params) {
+    const searchParams = page.params.map(urlParam => {
+      return urlParam.toSearchString({
+        value: targetParams[urlParam.name] ?? urlParam.getDefaultValue(),
+        mergeGlobals: false,
+        withPrefix: false,
+      });
+    });
 
-  if (params) {
-    return compile(extPageRoutePath)(params); // might throw error when required params not passed
+    if (searchParams.length > 0) {
+      stringifiedParams = `?${searchParams.join("&")}`;
+    }
   }
 
-  return extPageRoutePath;
+  return path.posix.join("/extension", sanitizeExtensionName(extensionId), pageId, stringifiedParams);
```

Comment:
Doesn't seem correct to use filepath join for urls.

Reply:
What is not correct?

Reply:
It's not safe for urls, see discussion: https://stackoverflow.com/questions/16301503/can-i-use-requirepath-join-to-safely-concatenate-urls

Reply:
Ok, will be fixed (but this PR already contained actually `path.join` for that reason)

Reply:
to be noted that when using `URL`, you need to provide base url in valid url format (ie http://...). Also it's not possible to join multiple paths, so in this case `path.posix.join` is IMHO valid.

Reply:
Just because `URL` doesn't work, it doesn't mean `path.posix.join` is correct. It probably works for our case so far, but could cause problems later.

There seems to be popular package `url-join` to do this safely.

We can change all path.posix.join's in one go in another PR if that's wanted though.